# Music Recommendation System (Top-10)

## Project Overview
This project builds a **music recommender system** that generates **top-10 song recommendations per user** using **implicit feedback** (listens, likes, skips, playlist adds).

The goal is to simulate a realistic recommendation pipeline:
- start from **messy interaction logs**
- perform **data cleaning and preprocessing**
- train a **collaborative filtering model**
- evaluate recommendations using **ranking-based metrics**

## Objective
For each user, recommend **10 songs** they are most likely to positively engage with next.

Formally: Given a user’s historical interactions, rank unseen songs and return the top 10.

## Dataset
- File: `music_interactions_messy.csv`
- Type: implicit user–song interaction data
- Characteristics:
  - duplicates
  - missing values
  - inconsistent labels
  - outliers
  - invalid timestamps

This dataset is intentionally **not clean** to reflect real-world data challenges. It is not a real dataset but one I created to simulate the process.
I felt that in the real world you are rarely given a clean dataset, so I decided to give myself a messy dataset to work with.

## Modeling Approach
- Feedback type: **Implicit**
- Primary model: **Matrix Factorization (ALS)**
- Library: `implicit`
- Recommendation type: **Top-N ranking (N = 10)**

## Evaluation
Offline evaluation using:
- **Recall@10**
- **Precision@10**
- **NDCG@10**

Baseline comparison:
- Global popularity-based recommender

## Deliverables
- Cleaned interaction dataset
- Trained recommender model
- Offline evaluation results
- Top-10 recommendations per user produced during evaluation



In [14]:
# Core
import numpy as np
import pandas as pd

# Visualization (optional but useful for EDA)
import matplotlib.pyplot as plt
import seaborn as sns

# Sparse matrices
from scipy.sparse import coo_matrix, csr_matrix

# Recommender models
from implicit.als import AlternatingLeastSquares

# Evaluation helpers
from sklearn.preprocessing import MinMaxScaler

# Utilities
import warnings
warnings.filterwarnings("ignore")

# Reproducibility
RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

# Display settings
pd.set_option("display.max_columns", None)
pd.set_option("display.width", 120)


## 1. Environment Setup and Imports

This section prepares the Python environment for the recommender system project.
No data is loaded and no modeling is performed here; this cell only configures tools and settings that will be used throughout the notebook.

### Library Imports

- **NumPy (`numpy`)**
  Used for numerical operations, random number generation, and efficient array handling.

- **Pandas (`pandas`)**
  Used for loading, cleaning, and manipulating tabular data such as the user–song interaction logs.

- **Matplotlib & Seaborn**
  Visualization libraries used during exploratory data analysis (EDA) to inspect distributions, missing values, and outliers.

- **SciPy Sparse Matrices**
  Recommender systems operate on large, mostly-empty user–item matrices.
  Sparse matrix formats (`COO`, `CSR`) allow efficient storage and computation.

- **Implicit ALS (`implicit.als`)**
  Provides the Alternating Least Squares matrix factorization model optimized for **implicit feedback** data.

- **Scikit-learn utilities**
  Used later for scaling values and supporting evaluation tasks.

### Warning Management

```python
warnings.filterwarnings("ignore")
```

### Reproducibility

```python
RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)
```

Setting a fixed random seed ensures:
- consistent train/test splits
- repeatable model initialization
- reproducible evaluation results
- This is essential for debugging and fair model comparison.

### Display Configuration

```python
pd.set_option("display.max_columns", None)
pd.set_option("display.width", 120)
```

Configures how DataFrames are displayed in the notebook:
- all columns are shown
- tables are wider and easier to read
These settings **do not modify the data itself.**

### Summary

This setup cell initializes all dependencies and global settings required for the project.
Once this cell runs successfully, the notebook is ready to load and inspect the dataset.



In [15]:
# Load the dataset
df = pd.read_csv("music_interactions_messy.csv")

# Basic inspection
print("Shape of dataset:", df.shape)
display(df.head())

# Column info and data types
df.info()

Shape of dataset: (10000, 13)


,user_id,song_id,event_type,play_count,liked,added_to_playlist,skipped,listen_seconds,timestamp,device,country,artist_id,genre
0,u00948,s000072,play,1,0,0,0,28,2025-06-13 13:57:14,mobile,US,a00189,pop
1,u00804,s000015,ADD_TO_PLAYLIST,6,1,0,0,332,2025-05-04 14:24:03,mobile,US,a00156,edm
2,u00570,s000365,like,17,1,0,0,289,2025-08-20 07:16:49,mobile,US,a00128,edm
3,u00824,s000024,ADD_TO_PLAYLIST,9,1,0,0,184,2025-09-25 18:22:14,mobile,US,a00065,latin
4,u00688,s000001,PLAY,14,0,0,0,342,2025-02-05 04:49:01,mobile,US,a00197,indie


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   user_id            10000 non-null  object
 1   song_id            10000 non-null  object
 2   event_type         10000 non-null  object
 3   play_count         10000 non-null  int64 
 4   liked              10000 non-null  int64 
 5   added_to_playlist  10000 non-null  int64 
 6   skipped            10000 non-null  int64 
 7   listen_seconds     10000 non-null  int64 
 8   timestamp          10000 non-null  object
 9   device             10000 non-null  object
 10  country            10000 non-null  object
 11  artist_id          10000 non-null  object
 12  genre              10000 non-null  object
dtypes: int64(5), object(8)
memory usage: 1015.8+ KB


## 2. Initial Data Inspection

After loading the dataset, we perform a high-level inspection to understand its structure, size, and obvious quality issues.


### Dataset Shape
The dataset contains a large number of user-to-song interaction records, with each row representing a single interaction event between a user and a song.

This confirms that the data is in a **long format**, which is appropriate for recommender systems.

### Columns Overview
From the preview and `df.info()` output, we observe the following types of variables:

- **Identifiers**
  - `user_id`
  - `song_id`
  - `artist_id`

- **Interaction Signals**
  - `event_type` (e.g., play, like, skip)
  - `play_count`
  - `liked`
  - `added_to_playlist`
  - `skipped`
  - `listen_seconds`

- **Contextual Metadata**
  - `timestamp`
  - `device`
  - `country`
  - `genre`

This structure provides enough information to construct an implicit feedback signal for recommendation.


### Data Quality Issues Observed
Based on the initial inspection:

- Several columns contain **missing or null values**
- `play_count` is not consistently numeric
- `event_type` uses **inconsistent labels and casing**
- `timestamp` includes invalid or non-standard values
- Categorical fields (e.g., `device`, `genre`) contain inconsistent formatting
- Duplicate rows may be present

These issues are expected and intentional, as the dataset is designed to simulate real-world logging data.


### Implications for Modeling
Before building a recommender model, the dataset will require:
- cleaning and standardization
- removal or correction of invalid records
- conversion into a numerical interaction strength
- filtering of sparse users and items

The next section focuses on **systematically identifying and addressing these data quality problems**.



In [29]:
# Basic statistics for numeric columns
display(df.describe(include=[np.number]))

# Unique value inspection for key categorical columns
for col in ["event_type", "device", "genre"]:
    print(f"\nUnique values in {col}:")
    print(df[col].value_counts().head(15))

# Check play_count type issues
print("\nNon-numeric play_count values:")
display(df[pd.to_numeric(df["play_count"], errors="coerce").isna()][["play_count"]].head())

# Timestamp parsing issues
parsed_ts = pd.to_datetime(df["timestamp"], errors="coerce")
print("\nInvalid timestamps:", parsed_ts.isna().sum())

# Unknown / blank user or song IDs
bad_users = df["user_id"].astype(str).str.strip().isin(["", "nan", "None", "guest"])
bad_songs = df["song_id"].astype(str).str.strip().isin(["", "nan", "None", "UNKNOWN"])

print("Rows with invalid user_id:", bad_users.sum())
print("Rows with invalid song_id:", bad_songs.sum())


,play_count,liked,added_to_playlist,skipped,listen_seconds
count,10000.000000,10000.000000,10000.0,10000.000000,10000.000000
mean,10.115500,0.409500,0.0,0.105300,244.621700
std,5.724966,0.491766,0.0,0.306955,113.786554
min,1.000000,0.000000,0.0,0.000000,5.000000
25%,6.000000,0.000000,0.0,0.000000,201.000000
50%,11.000000,0.000000,0.0,0.000000,267.000000
75%,15.000000,1.000000,0.0,0.000000,333.000000
max,19.000000,1.000000,0.0,1.000000,399.000000



Unique values in event_type:
event_type
PLAY               2439
play               2413
like               1051
LIKE               1033
ADD_TO_PLAYLIST    1022
add_to_playlist     989
SKIP                550
skip                503
Name: count, dtype: int64

Unique values in device:
device
mobile    10000
Name: count, dtype: int64

Unique values in genre:
genre
hip-hop      1450
rock         1333
latin        1261
classical    1221
indie        1199
edm          1193
pop          1177
jazz         1166
Name: count, dtype: int64

Non-numeric play_count values:


,play_count



Invalid timestamps: 0
Rows with invalid user_id: 0
Rows with invalid song_id: 0


## 3. Data Audit and Quality Assessment

Before performing any cleaning or preprocessing, we conduct a systematic audit of the dataset to identify data quality issues that could impact modeling and evaluation.

This step focuses on **measuring and confirming problems**, not fixing them yet.


### Inconsistent and Noisy Labels
Categorical columns exhibit inconsistent formatting:
- `event_type` contains multiple variants of the same action (different casing or alternative labels)
- `device` and `genre` values show inconsistent capitalization and formatting

These inconsistencies prevent reliable grouping and aggregation and will be standardized in the next section.



### Data Type Issues and Outliers
- The `play_count` column contains non-numeric values and extreme outliers
- `listen_seconds` includes negative values and unusually large durations

These issues suggest logging errors and require conversion, correction, or clipping to reasonable ranges.



### Invalid Timestamps
Some rows contain timestamps that cannot be parsed into valid datetime objects.
Since time-based splitting is required for realistic evaluation, invalid timestamps must be handled or removed.



### Invalid or Unknown Identifiers
The audit also identifies rows with:
- blank or placeholder `user_id` values
- unknown or invalid `song_id` values

Such rows cannot be used for collaborative filtering and will be removed during preprocessing.



### Summary
This audit confirms that the dataset reflects realistic, noisy interaction logs.
The next section focuses on **cleaning and standardizing the data** so it can be transformed into a numerical implicit feedback representation suitable for recommendation modeling.



In [18]:
# Make a working copy
df_clean = df.copy()

# Standardize IDs
for col in ["user_id", "song_id", "artist_id"]:
    df_clean[col] = (df_clean[col]
        .astype(str)
        .str.strip()
        .str.lower()
        .replace({"": np.nan, "none": np.nan, "nan": np.nan}))

# Standardize event_type labels
event_map = {
    "play": "play",
    "listen": "play",
    "replay": "replay",
    "repeat": "replay",
    "like": "like",
    "thumbs_up": "like",
    "skip": "skip",
    "skipped": "skip",
    "addtoplaylist": "add_to_playlist",
    "add2playlist": "add_to_playlist",
    "add_to_playlist": "add_to_playlist",
    "share": "share"}

df_clean['event_type'] = df_clean['event_type'].str.lower().str.strip()

df_clean["event_type"] = (
    df_clean["event_type"]
    .astype(str)
    .str.strip()
    .str.lower()
    .map(event_map))

# Coerce play_count to numeric
df_clean["play_count"] = pd.to_numeric(
    df_clean["play_count"], errors="coerce")

# Fix listen_seconds
df_clean["listen_seconds"] = pd.to_numeric(
    df_clean["listen_seconds"], errors="coerce")

# Remove negative durations
df_clean.loc[df_clean["listen_seconds"] < 0, "listen_seconds"] = np.nan

# Parse timestamps
df_clean["timestamp"] = pd.to_datetime(
    df_clean["timestamp"], errors="coerce")

# Drop rows with invalid core identifiers
df_clean = df_clean.dropna(subset=["user_id", "song_id", "timestamp"])

# Drop exact duplicates
df_clean = df_clean.drop_duplicates()

print("Shape before cleaning:", df.shape)
print("Shape after cleaning:", df_clean.shape)


Shape before cleaning: (10000, 13)
Shape after cleaning: (10000, 13)


## 4. Data Cleaning and Standardization

In this section, the dataset is cleaned and standardized to prepare it for modeling.

Based on the issues identified during the data audit, the following steps are performed:

- **Identifier normalization**
  User, song, and artist IDs are converted to a consistent format by trimming whitespace, standardizing case, and handling invalid or missing values.

- **Event type standardization**
  Inconsistent interaction labels are mapped to a fixed set of main event types (e.g., `play`, `like`, `skip`, `add_to_playlist`), enabling reliable aggregation and weighting.

- **Type coercion and validation**
  Numeric columns such as `play_count` and `listen_seconds` are safely converted to numeric types, with invalid entries handled appropriately.

- **Timestamp parsing**
  Interaction timestamps are parsed into datetime objects, with invalid timestamps removed to allow time-based splitting later in the pipeline.

- **Duplicate and invalid row removal**
  Rows with missing core identifiers or duplicate interactions are dropped to prevent bias in the recommender model.

After this step, the dataset represents a clean, consistent interaction log suitable for constructing an implicit feedback signal.

**Note**: Since the data is synthetic, the cleaning process is intentionally straightforward, but it reflects typical steps applied to messy interaction logs.



In [19]:
df_clean.head()

,user_id,song_id,event_type,play_count,liked,added_to_playlist,skipped,listen_seconds,timestamp,device,country,artist_id,genre
0,u00948,s000072,play,1,0,0,0,28.0,2025-06-13 13:57:14,mobile,US,a00189,pop
1,u00804,s000015,add_to_playlist,6,1,0,0,332.0,2025-05-04 14:24:03,mobile,US,a00156,edm
2,u00570,s000365,like,17,1,0,0,289.0,2025-08-20 07:16:49,mobile,US,a00128,edm
3,u00824,s000024,add_to_playlist,9,1,0,0,184.0,2025-09-25 18:22:14,mobile,US,a00065,latin
4,u00688,s000001,play,14,0,0,0,342.0,2025-02-05 04:49:01,mobile,US,a00197,indie


In [20]:
# Create a copy for interaction engineering
df_interactions = df_clean.copy()

# 1. Base strength from play_count
# We don't clip at 10 immediately; let super-fans have higher weight
df_interactions["play_strength"] = df_interactions["play_count"].fillna(1)

# 2. Event-based boosting (The Signal)
# Give massive boosts to explicit positive feedback
df_interactions["event_strength"] = 0.0

# If they liked/added to playlist, we treat this as equivalent to many plays
df_interactions.loc[df_interactions["event_type"] == "like", "event_strength"] += 5.0
df_interactions.loc[df_interactions["event_type"] == "add_to_playlist", "event_strength"] += 4.0
df_interactions.loc[df_interactions["event_type"] == "replay", "event_strength"] += 2.0

# 3. Handle Skips (The Noise)
# Instead of subtracting, we will set their strength to near zero later
is_skipped = (df_interactions["event_type"] == "skip") | (df_interactions["skipped"] == 1)

# 4. Calculate Final Strength
df_interactions["strength"] = df_interactions["play_strength"] + df_interactions["event_strength"]

# 5. Apply Skip Penalty
# If a song was skipped, zero out the strength (implicit ALS ignores 0s)
df_interactions.loc[is_skipped, "strength"] = 0.0

# Keep only what we need
df_interactions = df_interactions[["user_id", "song_id", "artist_id", "timestamp", "strength"]]

# Remove 0 strength interactions (saves memory and noise)
df_interactions = df_interactions[df_interactions["strength"] > 0]

display(df_interactions.head())

,user_id,song_id,artist_id,timestamp,strength
0,u00948,s000072,a00189,2025-06-13 13:57:14,1.0
1,u00804,s000015,a00156,2025-05-04 14:24:03,10.0
2,u00570,s000365,a00128,2025-08-20 07:16:49,22.0
3,u00824,s000024,a00065,2025-09-25 18:22:14,13.0
4,u00688,s000001,a00197,2025-02-05 04:49:01,14.0


## 5. Constructing Implicit Interaction Strength

This section converts cleaned interaction logs into a single **numeric implicit feedback signal** that can be used by recommender system models.

Collaborative filtering models do not operate on raw events (e.g., "play", "like", "skip").
Instead, each user-to-song interaction is represented by **one non-negative number** indicating the strength of user preference.


### Step 1: Create a Working Copy
A copy of the cleaned dataset is created so feature engineering does not modify the original cleaned data.


### Step 2: Base Strength from Play Count
The number of times a user played a song (`play_count`) is used as the base interaction signal.

- Missing `play_count` values are treated as `1`, meaning at least one interaction occurred.
- No manual capping is applied in the code; high play counts are kept to reflect strong preference.

This encodes the assumption: more plays imply stronger interest.


### Step 3: Initialize Event-Based Adjustment
An event adjustment column is initialized at zero for all rows.
This column is used to reward or penalize interactions based on explicit actions.


### Step 4: Reward Positive Actions
Certain user actions provide stronger evidence of preference and are rewarded:

- **Like** events add a strong positive signal
- **Add to playlist** events add a positive signal
- **Replay** events add a smaller positive signal

These actions are translated into numeric bonuses to increase interaction strength.

### Step 5: Handle Skips
Skips are treated as negative feedback.
Instead of subtracting, any skipped interaction is set to **zero strength**, which ALS effectively ignores.

This allows the model to distinguish between:
- songs a user actively engaged with
- songs a user skipped


### Step 6: Combine Signals into Final Strength
The final interaction strength is computed as:

#### play strength + event-based adjustment


### Step 7: Enforce Non-Negativity and Remove Zeros
All strengths are non-negative, and zero-strength rows are removed to reduce noise.


### Step 8: Select Modeling Columns
Only the columns required for downstream modeling are retained:
- user ID
- song ID
- artist ID
- timestamp
- interaction strength

The resulting table represents a clean, numerical interaction log suitable for constructing a user-to-item matrix and training a recommender system.


### Output Interpretation
Each row now represents:

*User U expressed an implicit preference of strength S for song I at time T.*

This format is the direct input required for collaborative filtering models such as implicit ALS.



In [21]:
# Filter sparse users and songs

min_user_interactions = 5
min_song_interactions = 5

# Count interactions per user and per song
user_counts = df_interactions["user_id"].value_counts()
song_counts = df_interactions["song_id"].value_counts()

# Keep only users/songs that meet the minimum thresholds
keep_users = user_counts[user_counts >= min_user_interactions].index
keep_songs = song_counts[song_counts >= min_song_interactions].index

df_filtered = df_interactions[
    df_interactions["user_id"].isin(keep_users) &
    df_interactions["song_id"].isin(keep_songs)].copy()

print("Before filtering:", df_interactions.shape)
print("After filtering: ", df_filtered.shape)
print("Unique users:", df_filtered["user_id"].nunique())
print("Unique songs:", df_filtered["song_id"].nunique())


Before filtering: (8947, 5)
After filtering:  (8438, 5)
Unique users: 943
Unique songs: 402


## 6. Filtering Sparse Users and Songs

After constructing the implicit interaction strength, the dataset is filtered to remove users and songs with very few interactions.

Collaborative filtering models rely on shared interaction patterns across users and items.
Users or songs with extremely limited interaction history do not provide enough information for learning meaningful representations.


### User Filtering
Users with fewer than a minimum number of interactions are removed.

- Very sparse users introduce noise
- Their preferences cannot be reliably inferred
- They can negatively affect evaluation metrics


### Song Filtering
Songs with very few interactions are also removed.

- Sparse songs lack enough data to position them correctly in latent space
- Recommendations involving such items tend to be unstable



### Thresholds
Minimum thresholds are applied to both users and songs:
- **Users:** at least 5 interactions
- **Songs:** at least 5 interactions

These values balance data retention with model stability and can be adjusted based on dataset size.



### Result
After filtering:
- the dataset contains only users and songs with sufficient interaction history
- the remaining interactions are better suited for collaborative filtering
- model training and evaluation become more reliable

This filtered interaction table serves as the input for ID encoding and user-to-item matrix construction in the next section.



In [22]:
# Encode IDs and build sparse user-item matrix

# Create integer index mappings
user_ids = df_filtered["user_id"].unique()
song_ids = df_filtered["song_id"].unique()

user_to_idx = {u: i for i, u in enumerate(user_ids)}
song_to_idx = {s: i for i, s in enumerate(song_ids)}

# Map to indices
df_filtered["user_idx"] = df_filtered["user_id"].map(user_to_idx)
df_filtered["song_idx"] = df_filtered["song_id"].map(song_to_idx)

# Build sparse matrix (users x songs)
rows = df_filtered["user_idx"].to_numpy()
cols = df_filtered["song_idx"].to_numpy()
data = df_filtered["strength"].to_numpy()

user_item_matrix = coo_matrix(
    (data, (rows, cols)),
    shape=(len(user_ids), len(song_ids))).tocsr()

print(user_item_matrix)

print("User-item matrix shape:", user_item_matrix.shape)
print("Non-zero entries:", user_item_matrix.nnz)
print("Sparsity:", 1 - (user_item_matrix.nnz / (user_item_matrix.shape[0] * user_item_matrix.shape[1])))


<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 6030 stored elements and shape (943, 402)>
  Coords	Values
  (0, 0)	1.0
  (0, 1)	5.0
  (0, 44)	13.0
  (0, 233)	18.0
  (0, 329)	1.0
  (1, 1)	27.0
  (1, 33)	12.0
  (1, 38)	7.0
  (1, 44)	10.0
  (1, 89)	36.0
  (1, 165)	13.0
  (1, 211)	17.0
  (1, 274)	1.0
  (2, 2)	22.0
  (2, 33)	11.0
  (2, 38)	7.0
  (2, 44)	44.0
  (2, 274)	1.0
  (3, 3)	66.0
  (3, 50)	17.0
  (3, 54)	13.0
  (3, 168)	11.0
  (3, 175)	46.0
  (3, 197)	5.0
  (3, 315)	1.0
  :	:
  (938, 4)	19.0
  (938, 8)	14.0
  (938, 46)	27.0
  (938, 83)	12.0
  (938, 251)	13.0
  (938, 272)	24.0
  (938, 358)	5.0
  (939, 35)	19.0
  (939, 59)	20.0
  (939, 148)	23.0
  (939, 277)	1.0
  (939, 285)	23.0
  (940, 5)	22.0
  (940, 50)	18.0
  (940, 177)	11.0
  (940, 312)	16.0
  (940, 369)	1.0
  (940, 386)	1.0
  (941, 7)	13.0
  (941, 51)	23.0
  (941, 65)	21.0
  (942, 6)	24.0
  (942, 40)	19.0
  (942, 298)	24.0
  (942, 324)	12.0
User-item matrix shape: (943, 402)
Non-zero entries: 6030
Sparsity: 0.9840

## 7. Encoding Identifiers and Constructing the User-to-Item Matrix

Recommender system models such as matrix factorization do not operate on raw string identifiers (e.g., user IDs or song IDs).
Instead, they learn from a **numerical matrix** where rows represent users, columns represent songs, and values represent interaction strength.

This section converts the interaction table into that required format.


### Why Encoding Is Necessary

- User and song identifiers are strings (e.g., "u12345", "s67890")
- Matrix-based models require **integer indices** starting from 0
- Each user must correspond to a row index
- Each song must correspond to a column index

Encoding ensures a **one-to-one mapping** between:
- a user and a matrix row
- a song and a matrix column


### Step 1: Create Index Mappings

Unique user IDs and song IDs are extracted from the filtered dataset and mapped to integers.

- `user_to_idx`: maps each user ID -> row index
- `song_to_idx`: maps each song ID -> column index

These mappings are stored so the model’s numeric outputs can later be translated back into real user and song IDs.


### Step 2: Map Identifiers to Indices

Each interaction row is updated with:
- `user_idx`: integer index representing the user
- `song_idx`: integer index representing the song

At this point, every interaction is fully numeric and ready for matrix construction.


### Step 3: Construct the Sparse User-to-Item Matrix

A **sparse matrix** is created where:
- rows = users
- columns = songs
- values = implicit interaction strength

Only observed interactions are stored.
Unobserved user-to-song pairs are implicitly treated as zero.

This is critical because recommender datasets are extremely sparse:
- most users interact with only a small fraction of all songs
- storing a dense matrix would be inefficient and unnecessary


### Step 4: Convert to CSR Format

The matrix is converted to **Compressed Sparse Row (CSR)** format.

CSR format:
- is memory-efficient
- supports fast row-based operations
- is required by the ALS implementation used later

### Matrix Interpretation

Each entry in the matrix represents the inferred strength of a user’s preference for a song.

- larger values -> stronger preference
- zero values -> no observed interaction

This matrix is the **core input** to the recommender model.

---

### Summary

This step bridges the gap between:
- real-world interaction logs and mathematical models used for recommendation

With the user-to-item matrix constructed, the data is now in the correct form for training a collaborative filtering model in the next section.



In [23]:
# Time-based train/test split (leave-last-1 per user)

df_time = df_filtered.sort_values(["user_idx", "timestamp"]).copy()

# For each user, hold out their last interaction as test
df_time["rank_within_user"] = df_time.groupby("user_idx").cumcount()
df_time["max_rank"] = df_time.groupby("user_idx")["rank_within_user"].transform("max")

test_df = df_time[df_time["rank_within_user"] == df_time["max_rank"]].copy()
train_df = df_time[df_time["rank_within_user"] < df_time["max_rank"]].copy()

print("Train interactions:", train_df.shape[0])
print("Test interactions:", test_df.shape[0])
print("Users in train:", train_df["user_idx"].nunique())
print("Users in test:", test_df["user_idx"].nunique())

# Build train matrix (users x songs)
train_matrix = coo_matrix(
    (train_df["strength"].to_numpy(),
     (train_df["user_idx"].to_numpy(), train_df["song_idx"].to_numpy())),
    shape=user_item_matrix.shape).tocsr()

# Build test dictionary: user -> set of held-out song indices
test_items = test_df.groupby("user_idx")["song_idx"].apply(set).to_dict()

train_matrix


Train interactions: 7495
Test interactions: 943
Users in train: 943
Users in test: 943


<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 5505 stored elements and shape (943, 402)>

## 8. Time-Based Train/Test Split

In this section, the interaction data is split into training and testing sets using a **time-based strategy**.
This approach is critical for recommender systems because it mirrors how recommendations are used in practice.

Rather than randomly splitting interactions, the model is trained on **past behavior** and evaluated on **future behavior**.


### Why Time-Based Splitting Is Necessary

In real-world recommendation settings:
- users interact with items over time
- recommendations are made based on historical interactions
- future interactions are unknown at training time

A random split would allow the model to train on future interactions, leading to **data leakage** and overly optimistic evaluation results.


### Step 1: Chronological Ordering

Interactions are first sorted by:
- user index
- timestamp

This ensures each user’s interaction history is processed in the correct temporal order.



### Step 2: Rank Interactions Within Each User

Each interaction is assigned a rank representing its order in time for a given user.

- the earliest interaction has the lowest rank
- the most recent interaction has the highest rank

This ranking allows us to identify which interaction occurred last for each user.



### Step 3: Hold Out the Most Recent Interaction

For every user:
- the **most recent interaction** is held out as test data
- all earlier interactions are used for training

This setup simulates the task: “Given everything we know about a user up to now, can we recommend what they interact with next?”


### Step 4: Construct Training and Test Sets

Two datasets are created:
- **Training set**: all interactions except the most recent one per user
- **Test set**: exactly one held-out interaction per user

This guarantees:
- every user appears in both training and test data
- evaluation is consistent across users



### Step 5: Build the Training Matrix

The training interactions are converted into a **user-to-item sparse matrix**.

- rows correspond to users
- columns correspond to songs
- values represent implicit interaction strength

Only training interactions are included in this matrix, ensuring the model has no access to test data during learning.



### Step 6: Prepare Test Items for Evaluation

The test interactions are stored as a mapping from:
- user index -> set of held-out song indices
- In other words, for every user, we keep track of which song was removed from their interaction history so we can verify whether it appears in the model’s recommendations.

This structure is used during evaluation to check whether the model’s top-N recommendations contain the true future interaction.


### Interpretation

At evaluation time, the model is asked: “For this user, can we rank the held-out song among the top recommendations?”

This directly supports metrics such as:
- Recall@K
- Precision@K
- NDCG@K

---

### Summary

This time-based split:
- prevents data leakage
- produces realistic evaluation results
- aligns with real-world recommendation scenarios

With the train/test split complete, the dataset is now ready for baseline modeling and recommender training in the next section.



In [ ]:
# Popularity baseline and evaluation (THE JUICY PART)

K = 10  # top-10 recommendations

# 1. Compute global song popularity from training data
song_popularity = (
    train_df
    .groupby("song_idx")["strength"]
    .sum()
    .sort_values(ascending=False))

popular_songs = song_popularity.index.tolist()

# 2. Generate top-K popular recommendations for each user
def recommend_popular(user_idx, K=3):
    return popular_songs[:K]

# 3. Evaluate Recall@K and Precision@K
def recall_at_k(recommended, relevant):
    if not relevant:
        return 0.0
    return len(set(recommended) & relevant) / len(relevant)

def precision_at_k(recommended, relevant, K):
    return len(set(recommended) & relevant) / K

recalls = []
precisions = []

for user_idx, relevant_items in test_items.items():
    recs = recommend_popular(user_idx, K)
    recalls.append(recall_at_k(recs, relevant_items))
    precisions.append(precision_at_k(recs, relevant_items, K))

print(f"Popularity Recall@{K}:{np.mean(recalls)}")
print(f"Popularity Precision@{K}:{np.mean(precisions)}")

# This answers the question: “Does our model actually do better than recommending the most popular songs?”


Popularity Recall@10:0.17709437963944857
Popularity Precision@10:0.01770943796394486


## 9. Popularity Baseline and Offline Evaluation

Before training a personalized recommender model, a **baseline** is established to measure how well a simple, non-personalized approach performs.

The popularity baseline recommends the same songs to every user based solely on overall interaction frequency.
Any personalized model should outperform this baseline to be considered useful.


### Popularity-Based Recommendation Strategy

Song popularity is computed using the **training data only** by aggregating interaction strength across all users.

Songs are ranked from most to least popular, and the top 10 most popular songs are recommended to every user, regardless of their individual listening history.

This baseline ignores personalization entirely and serves as a minimum performance benchmark.


### Evaluation Setup

Evaluation is performed using a **time-based train/test split**:
- the model is trained on past interactions
- the most recent interaction per user is held out as test data

Each user receives exactly 10 recommendations, and performance is evaluated by comparing these recommendations to the user’s held-out future interaction.


### Metrics Used

Two ranking-based metrics are computed:

- **Recall@10**
  Measures how often the held-out song appears in the top 10 recommendations.

Recall@10 asks: “Did we guess the right hidden song in our top 10 guesses?”

- **Precision@10**
  Measures the proportion of recommended songs that match the held-out interaction.

Precision@10 asks: “Out of the 10 songs we recommended, how many were correct?”

These metrics reflect the model’s ability to correctly rank a user’s future interaction within a small recommendation list.


### Results Interpretation

The popularity baseline achieves low Recall@10 and Precision@10 values.

This outcome is expected and indicates that:
- most users do not listen to the same globally popular songs
- music preferences are highly individualized
- recommending popular items alone is insufficient for accurate prediction

Low baseline performance confirms that the recommendation task is non-trivial and that personalization is necessary.


### Importance of the Baseline

Establishing a popularity baseline:
- validates the evaluation pipeline
- provides a reference point for improvement
- ensures that future models add value beyond simple heuristics

All subsequent recommender models will be evaluated relative to this baseline.

---

### Next Step

With a baseline established, the next section introduces a **personalized collaborative filtering model** using matrix factorization to improve recommendation quality.



In [ ]:
# Train ALS recommender model
from implicit.als import AlternatingLeastSquares

# Hyperparameters
ALPHA_VAL = 40  # CRITICAL: Scales the confidence. 40 is a standard starting point.
FACTORS = 64
ITERATIONS = 50

als_model = AlternatingLeastSquares(
    factors=FACTORS,
    regularization=0.05,
    iterations=ITERATIONS,
    random_state=RANDOM_STATE
)

# Apply Alpha Scaling (Confidence Weighting)
# ALS expects: Confidence = 1 + alpha * r_ui (or just alpha * r_ui)
data_conf = (train_matrix * ALPHA_VAL).astype('double')

# Train on the scaled matrix
als_model.fit(data_conf)

print(f"Model trained with Alpha={ALPHA_VAL}, Factors={FACTORS}")
print("train_matrix shape:", data_conf.shape)
print("user_factors shape:", als_model.user_factors.shape)
print("item_factors shape:", als_model.item_factors.shape)

  0%|          | 0/50 [00:00<?, ?it/s]

Model trained with Alpha=40, Factors=64
train_matrix shape: (943, 402)
user_factors shape: (943, 64)
item_factors shape: (402, 64)


## 10. Training a Personalized Recommender with ALS

In this section, a **personalized recommender model** is trained using **Alternating Least Squares (ALS)**, a matrix factorization technique designed for implicit feedback data.

Unlike the popularity baseline, ALS learns individualized preferences for each user based on their historical interactions.


### Model Input

The model is trained on the **user-to-item interaction matrix** constructed from past interactions only.
Each matrix entry represents the inferred strength of a user’s preference for a song.

To match the requirements of the `implicit` library, the matrix is scaled for confidence (alpha weighting) and transposed so that:
- rows correspond to songs
- columns correspond to users


### Model Overview

ALS factorizes the interaction matrix into two low-dimensional matrices:
- one representing **user latent factors**
- one representing **song latent factors**

Each user and each song is represented by a vector in a shared latent space.
Preference scores are computed using the dot product between user and song vectors.


### Training Procedure

The model is initialized with:
- a fixed number of latent factors
- regularization to prevent overfitting
- a set number of training iterations
- an alpha value to scale confidence in observed interactions

During training, ALS alternates between:
- fixing song factors and optimizing user factors
- fixing user factors and optimizing song factors

This iterative process minimizes reconstruction error while accounting for implicit feedback.

---

### Outcome

After training:
- each user has a learned representation of their listening preferences
- each song has a learned representation of its characteristics
- the model can generate personalized rankings of songs for each user

This trained model will be evaluated in the next section using ranking-based metrics and compared against the popularity baseline.



In [ ]:
# Evaluate ALS model @K 

K = 10

def ndcg_at_k(recommended, relevant, K):
    dcg = 0.0
    for i, item in enumerate(recommended[:K]):
        if item in relevant:
            dcg += 1.0 / np.log2(i + 2)
    ideal_hits = min(len(relevant), K)
    idcg = sum(1.0 / np.log2(i + 2) for i in range(ideal_hits))
    return 0.0 if idcg == 0 else dcg / idcg

def recall_at_k(recommended, relevant):
    return 0.0 if len(relevant) == 0 else len(set(recommended) & relevant) / len(relevant)

def precision_at_k(recommended, relevant, K):
    return len(set(recommended) & relevant) / K

recalls, precisions, ndcgs = [], [], []

for user_idx, relevant_items in test_items.items():
    rec_items, rec_scores = als_model.recommend(
        userid=user_idx,
        user_items=train_matrix[user_idx],
        N=K,
        filter_already_liked_items=True)

    rec_items = list(rec_items)
    recalls.append(recall_at_k(rec_items, relevant_items))
    precisions.append(precision_at_k(rec_items, relevant_items, K))
    ndcgs.append(ndcg_at_k(rec_items, relevant_items, K))

print(f"ALS Recall@{K}:{np.mean(recalls)}")
print(f"ALS Precision@{K}:{np.mean(precisions)}")
print(f"ALS NDCG@{K}:{np.mean(ndcgs)}")


ALS Recall@10:0.2067868504772004
ALS Precision@10:0.020678685047720042
ALS NDCG@10:0.1289989649271788


## 11. Evaluating the ALS Model (Top-K Ranking Performance)

In this section, the trained ALS recommender model is evaluated using **ranking-based metrics**.  
The goal is to measure how well the model predicts a user’s future interaction by ranking relevant songs near the top of the recommendation list.

Evaluation is performed using **Top-K recommendations**, where \( K = 10 \).

---

### Evaluation Setup

For each user:
- the model generates a ranked list of the top 10 recommended songs
- songs the user has already interacted with in training are filtered out
- the recommendations are compared against the user’s held-out future interaction(s)

This simulates a realistic scenario where a recommender suggests new content based on past behavior.


### Metrics Used

Three standard ranking metrics are computed:

#### Recall@K
Recall@K measures whether the model successfully includes the user’s future interaction among the top-K recommendations.

- A score of 1 indicates the held-out song appears in the top-K list
- A score of 0 indicates it does not

Averaged across users, Recall@K reflects how often the model “gets the right answer.”


#### Precision@K
Precision@K measures how many of the recommended songs are actually relevant.

Since only a small number of relevant items exist per user (often just one), Precision@K values are expected to be low in next-item prediction tasks.


#### NDCG@K (Normalized Discounted Cumulative Gain)
NDCG@K accounts for **ranking position**.

- Relevant items appearing higher in the recommendation list receive more credit
- Correct recommendations ranked lower receive less credit

This metric captures not only *whether* the model recommends the correct song, but *how highly* it ranks it.


### Interpretation

Evaluating at \( K = 10 \) provides a less restrictive view of model performance than Top-3 evaluation.

- Improvement at higher K values indicates that the model is learning meaningful for user to song relationships
- Strong performance at larger K often precedes improvements at smaller K through tuning and re-ranking

These metrics are compared against the popularity baseline to assess whether personalization provides measurable benefit.


### Summary

This evaluation confirms whether the ALS model:
- improves upon non-personalized recommendations
- ranks relevant items higher than irrelevant ones
- learns useful latent representations from implicit feedback

The results of this section guide further model tuning and optimization.


## 12. Results Interpretation and Model Comparison (K = 10)

This section compares the performance of the **personalized ALS recommender** against a **non-personalized popularity baseline**, using Top-10 recommendations.

Evaluation is restricted to \( K = 10 \) to provide a more realistic assessment of model performance given the sparsity of the data and the difficulty of next-item prediction.

---

### Summary of Results

| Model        | Recall@10 | Precision@10 | NDCG@10 |
|-------------|-----------|--------------|---------|
| Popularity  | 0.1771    | 0.0177       | —       |
| ALS (MF)    | 0.2068    | 0.0207       | 0.1290  |

---

### Recall@10 Comparison

Recall@10 measures how often the user’s held-out future interaction appears within the top 10 recommendations.

- The **popularity baseline** achieves a Recall@10 of approximately **0.18**, meaning that about 18% of users had their next song appear among the globally most popular songs.
- The **ALS model** improves Recall@10 to approximately **0.21**, indicating that personalization allows the model to correctly surface the relevant song for more users.

This improvement demonstrates that ALS is learning **user-specific preferences** beyond global popularity trends.


### Precision@10 Comparison

Precision@10 measures the proportion of recommended songs that are relevant.

- Precision values are low for both models, which is expected because each user typically has only **one relevant test item**.
- The ALS model achieves a higher Precision@10 than the popularity baseline, reflecting a higher concentration of relevant items within the recommendation list.

Low absolute precision values are normal in recommender systems evaluated under a leave-one-out, next-item prediction setting.


### NDCG@10 Interpretation

NDCG@10 evaluates not only whether the relevant item appears in the recommendation list, but **how highly it is ranked**.

- The ALS model achieves an NDCG@10 of approximately **0.13**.
- This indicates that when the ALS model successfully recommends the correct song, it tends to rank it closer to the top of the list rather than near the bottom.

This result confirms that ALS is learning meaningful latent representations that improve ranking quality, not just recall.


### Why K = 10 Is Appropriate

Evaluating at K = 10 is well suited for this dataset because:
- the interaction data is sparse
- each user has very few held-out relevant items
- music recommendation is exploratory, where users typically browse multiple suggestions

Top-3 evaluation is significantly more restrictive and may obscure meaningful improvements that are visible at higher K values.


### Overall Interpretation

The ALS recommender consistently outperforms the popularity baseline across all reported metrics at K = 10.

These results indicate that:
- personalization provides measurable benefits
- the model successfully captures user to song interaction patterns
- matrix factorization is an effective approach for this implicit-feedback music recommendation task

---

### Conclusion

This comparison validates the use of a personalized recommender over a popularity-based approach.  
While absolute metric values remain modest due to data sparsity and task difficulty, the observed improvements confirm that ALS is learning meaningful structure from the interaction data.


## 13. Limitations and Future Work

While the ALS-based recommender demonstrates clear improvements over a popularity baseline, several limitations remain.

---

### Data Limitations
The interaction data is sparse and noisy by design, reflecting realistic logging behavior.  
Each user has a limited interaction history, and evaluation is based on a single held-out future interaction, which makes next-item prediction particularly challenging.


### Model Limitations
The ALS model relies solely on collaborative filtering and does not incorporate song metadata such as genre, artist, or audio features.  
Additionally, ALS struggles with cold-start users and songs that lack sufficient interaction history.


### Evaluation Limitations
All evaluation is performed offline using ranking-based metrics.  
While these metrics are informative, they do not fully capture real user behavior or satisfaction, which would require online experimentation such as A/B testing.

The choice of K=10 for evaluation provides a broader view of model capability, though the final product target remains a tight Top-3 recommendation set.


### Future Work
Potential improvements include:
- more sophisticated confidence weighting for implicit feedback
- hybrid models that combine collaborative filtering with content-based features
- neural recommendation models for capturing more complex user–item interactions
- post-processing and re-ranking techniques to improve diversity and reduce popularity bias

These directions provide clear opportunities to further enhance recommendation quality.


In [30]:
# Generate final recommendations and export to CSV

# 1. Create inverse mappings (Index -> ID)
idx_to_user = {i: u for u, i in user_to_idx.items()}
idx_to_song = {i: s for s, i in song_to_idx.items()}

# 2. Prepare metadata lookup for a readable CSV
# We take the unique song information from our cleaned dataframe
song_metadata = df_clean[['song_id', 'artist_id', 'genre']].drop_duplicates('song_id')

all_recs = []
N_RECS = 10

print(f"Generating Top-{N_RECS} recommendations for {len(user_ids)} users...")

# 3. Loop through every user and generate their Top-N
for user_id in user_ids:
    user_idx = user_to_idx[user_id]
    
    # Request recommendations from the model
    # filter_already_liked_items=True ensures we don't recommend what they've already played
    ids, scores = als_model.recommend(
        userid=user_idx, 
        user_items=train_matrix[user_idx], 
        N=N_RECS, 
        filter_already_liked_items=True
    )
    
    # Store each recommendation in a flat list
    for rank, (song_idx, score) in enumerate(zip(ids, scores), 1):
        all_recs.append({
            'user_id': user_id,
            'song_id': idx_to_song[song_idx],
            'score': float(score), # Conversion for compatibility
            'rank': rank
        })

# 4. Create the final DataFrame
recs_df = pd.DataFrame(all_recs)

# 5. Enrich with Metadata (Artist/Genre)
recs_df = recs_df.merge(song_metadata, on='song_id', how='left')

# 6. Save to CSV for the "usable format"
recs_df.to_csv("final_recommendations.csv", index=False)

print("Success! Recommendations exported to 'final_recommendations.csv'")

# Display a sample of the results (first 2 users)
display(recs_df.head(20))

Generating Top-10 recommendations for 943 users...
Success! Recommendations exported to 'final_recommendations.csv'


,user_id,song_id,score,rank,artist_id,genre
0,u00948,s000171,0.488470,1,a00031,jazz
1,u00948,s000391,0.478560,2,a00011,edm
2,u00948,s000316,0.476952,3,a00097,pop
3,u00948,s000494,0.441793,4,a00145,edm
4,u00948,s000025,0.398942,5,a00009,edm
5,u00948,s000330,0.382702,6,a00106,edm
6,u00948,s000360,0.380377,7,a00153,edm
7,u00948,s000232,0.363711,8,a00181,classical
8,u00948,s000161,0.363669,9,a00188,classical
9,u00948,s000140,0.362360,10,a00179,edm


## 14. Generating and Exporting Final Recommendations

In this section, the trained ALS model is used to generate **final Top-K music recommendations** for each user.  
This step converts the recommender from an experimental model into a usable system output.

---

### Recommendation Generation

For every user in the dataset:
- the model produces a ranked list of recommended songs
- songs the user has already interacted with are excluded
- recommendations are ranked by the model’s predicted preference score

### Exporting Final Recommendations

The resulting recommendations reflect personalized listening preferences learned from historical interactions.

This final section transitions the model from an experimental phase to a ***production-ready output.*** While the model evaluates performance using internal integer indices, a real-world application requires human-readable IDs and metadata (Artist, Genre) to be useful for a UI or marketing team.

***The Recommendation Pipeline***
- Inverse Mapping: We convert the internal user_idx and song_idx back to their original unique identifiers (e.g., u00123, s00456).
- Personalized Ranking: For every user in the dataset, the ALS model computes a preference score for all songs they have not previously interacted with.
- Metadata Enrichment: We join the recommended IDs with our song metadata table to include Artist ID and Genre. This allows for a "sanity check" to ensure the recommendations align with the user's known tastes.
- Data Persistence: The results are flattened and exported to `final_recommendations.csv` for use in external applications.


### Result

The final output provides a concrete, user-facing recommendation list that demonstrates the practical applicability of the model.  
This completes the end-to-end recommendation pipeline, from raw interaction data to deployable recommendations.

---

### Closing

Thank you for viewing this project and for your time.
